aten/src/ATen/NamedTensorUtils.h

```diff
@@ -43,26 +43,97 @@ unify_from_right(DimnameList names, DimnameList other, const char* action = "bro
 
 namespace namedinference {
 
-// Names get propagated via the following rules:
-// 1) If result does not have names, then `names` get propagated.
-// 2) If result has names, then `names` must be equal to result.names
-void propagate_names(Tensor& result, optional<DimnameList> names);
-void propagate_names(Tensor& result, std::vector<Dimname>&& names, bool validate_names);
-CAFFE2_API void propagate_names(Tensor& result, optional<std::vector<Dimname>>&& maybe_names, bool validate_names);
-void propagate_names(TensorImpl* result, optional<DimnameList> names);
-void propagate_names(TensorImpl* result, std::vector<Dimname>&& names, bool validate_names);
-void propagate_names(TensorImpl* result, optional<std::vector<Dimname>>&& maybe_names, bool validate_names);
+// [NOTE] Writing name inference rules
+//
+// Operators that support named tensors generally look like the following:
+//
+// Tensor op(...) {
+//   perform_shape_checks(...);
+//   # (1)
+//   auto maybe_outnames = compute_outnames(...);
+//   auto result = [&]() {
+//     NoNamesGuard guard;
+//     return op_impl(...);
+//   }();
+//   # (2)
+//   propagate_names_if_nonempty(result, maybe_outnames);
+//
+// Each op has (1) a compute outnames step and (2) a propagate names step.
+//
+// compute_outnames is responsible for checking that input names match and
+// determining what the output names should be. It returns either:
+// - {} (if the inputs tensors are all unnamed)
+// - non-empty outnames.
+//
+// propagate_names_if_nonempty propagates the outnames if they exist to the result
+// tensors.
+//
+// The {} case is an optimization; if the user does not use named tensors they
+// pay no perf cost for it.
+
+// Propagates `names` to `result` if `names` is not empty.
+// `names` can be empty; see [NOTE] Writing name inference rules
```

Comment:
this `see [NOTE]` seems recursive

Reply:
There's a space between the [NOTE] and this block that mentions the note. This block is supposed to be documentation for `propagate_names_if_nonempty`.

Reply:
I'll add an extra space to make it clearer

Reply:
ah I see. some other ideas: you could start/end the note with a line of `////////////`'s or move the note above the start of the namespace, or use /**  */'s to better distinguish it from comments on the specific function below.